docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,838 @@
+    +   if (pargs_in->ptr && pargs_in->ptr->blob_array)
+    +       OE_ADD_SIZE(_deepcopy_out_buffer_size, ((size_t)pargs_in->ptr->num * sizeof(Blob)));
+    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr->num; _i_2++)
+    +   {
+    +       if (pargs_in->ptr && pargs_in->ptr->blob_array && pargs_in->ptr->blob_array[_i_2].buf)
+    +           OE_ADD_SIZE(_deepcopy_out_buffer_size, pargs_in->ptr->blob_array[_i_2].len);
+    +   }
+
+    +   if (_deepcopy_out_buffer_size)
+    +   {
+    +       _deepcopy_out_buffer = (uint8_t*) malloc(_deepcopy_out_buffer_size);
+    +       if (!_deepcopy_out_buffer)
+    +       {
+    +           _result = OE_OUT_OF_MEMORY;
+    +           goto done;
+    +       }
+    +   }
+
+    +   /* Serialize the deep-copied content into the buffer. */
+    +   if (pargs_in->ptr && pargs_in->ptr->blob_array)
+    +       OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr->blob_array, ((size_t)pargs_in->ptr->num * sizeof(Blob)));
+    +   for (size_t _i_2 = 0; _i_2 < pargs_in->ptr->num; _i_2++)
+    +   {
+    +       if (pargs_in->ptr && pargs_in->ptr->blob_array && pargs_in->ptr->blob_array[_i_2].buf)
+    +           OE_WRITE_DEEPCOPY_OUT_PARAM(pargs_in->ptr->blob_array[_i_2].buf, pargs_in->ptr->blob_array[_i_2].len);
+    +   }
+
+    +   if (_deepcopy_out_buffer_offset != _deepcopy_out_buffer_size)
+    +   {
+    +       _result = OE_FAILURE;
+    +       goto done;
+    +   }
+    +   OE_ADD_SIZE(output_buffer_offset, _deepcopy_out_buffer_size);
+
+    +   /* Pass the pointer of the deep-copied content to the enclave. */
+    +   pargs_out->_host_buffer = _deepcopy_out_buffer;
+
+        /* Propagate errno back to enclave. */
+        /* Errno propagation not enabled. */
+
+        /* Success. */
+        _result = OE_OK;
+        *output_bytes_written = output_buffer_offset;
+
+    done:
+    +   /* Free pargs_out->_host_buffer on failure. */
+    +   if (_result != OE_OK)
+    +   {
+    +       free(pargs_out->_host_buffer);
+    +       pargs_out->_host_buffer = NULL;
+    +   }
+
+    +   /* Free nested buffers allocated by the host. */
+    +   if (pargs_in->ptr)
+    +   {
+    +       for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
+    +       {
+    +           if (pargs_in->ptr[_i_1].blob_array)
+    +           {
+    +               for (size_t _i_2 = 0; _i_2 < pargs_in->ptr[_i_1].num; _i_2++)
+    +               {
+    +                   free(pargs_in->ptr[_i_1].blob_array[_i_2].buf);
+    +                   pargs_in->ptr[_i_1].blob_array[_i_2].buf = NULL;
+    +               }
+    +           }
+    +           free(pargs_in->ptr[_i_1].blob_array);
+    +           pargs_in->ptr[_i_1].blob_array = NULL;
+    +       }
+    +   }
+
+        if (pargs_out && output_buffer_size >= sizeof(*pargs_out))
+            pargs_out->_result = _result;
+    }
+    ```
+
+Discussion
+-----------
+
+The proposed design relies on the shared memory between an enclave and a host and therefore does not apply
```

Comment:
By moving host buffer allocation to the infrastructure rather than the generated code, there is no longer any reliance on edger8r generated code on shared memory. It is the infrastructure that needs to marshal the deep-copy parameter appropriately on each platform (SGX, OPTEE).

Reply:
Removed the section.